### PR TITLE
Scripts flattened into a different namespace

### LINKED ISSUE #167: Scripts to be extensions in different namespace
# Scripts namespace change

All Scripts (Stored Procedures, UDFs, Triggers) will move to the `Microsoft.Azure.Cosmos.Scripts` namespace and inlined.

Based on a Container, the Scripts can be accessed by calling the extension `GetScripts()`.

For example

    CosmosScripts cosmosScripts = container.GetScripts();
    CosmosStoredProcedureSettings storedProcedure = await cosmosScripts.CreateStoredProcedureAsync("scriptId", script);
    string result = await cosmosScripts.ExecuteStoredProcedureAsync<object ,string >(id : "scriptId", partitionKey : documentDefinition.Id, input : null);

### PR DESCRIPTION
**Pull request by ealsur**
# Scripts namespace change

## Description
All Scripts (Stored Procedures, UDFs, Triggers) will move to the `Microsoft.Azure.Cosmos.Scripts` namespace.

Based on a Container, the Scripts can be accessed by calling the extension `GetScripts()`.

For example:

    CosmosScripts cosmosScripts = container.GetScripts();
    CosmosStoredProcedureSettings storedProcedure = await cosmosScripts.CreateStoredProcedureAsync("scriptId", script);
    string result = await cosmosScripts.ExecuteStoredProcedureAsync<object ,string >(id : "scriptId", partitionKey : documentDefinition.Id, input : null);

This includes all script operations:

* cosmosScripts.CreateStoredProcedureAsync
* cosmosScripts.GetStoredProcedureIterator
* cosmosScripts.ReadStoredProcedureAsync
* cosmosScripts.ReplaceStoredProcedureAsync
* cosmosScripts.DeleteStoredProcedureAsync
* cosmosScripts.ExecuteStoredProcedureAsync
* cosmosScripts.CreateTriggerAsync
* cosmosScripts.GetTriggerIterator
* cosmosScripts.GetTriggerIterator
* cosmosScripts.ReplaceTriggerAsync
* cosmosScripts.DeleteTriggerAsync
* cosmosScripts.CreateUserDefinedFunctionAsync
* cosmosScripts.GetUserDefinedFunctionIterator
* cosmosScripts.ReadUserDefinedFunctionAsync
* cosmosScripts.ReplaceUserDefinedFunctionAsync
* cosmosScripts.DeleteUserDefinedFunctionAsync

## Type of change

Please delete options that are not relevant.

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Closing issues

Fixes #167 .

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Resource/CosmosResponseFactory.cs [hunk 1]
```diff
@@ -72,48 +73,51 @@ internal CosmosResponseFactory(CosmosJsonSerializer cosmosJsonSerializer)
...
+            {
+                T item = this.ToObjectInternal<T>(cosmosResponseMessage);
+                return new CosmosStoredProcedureExecuteResponse<T>(
+                    cosmosResponseMessage.StatusCode,
+                    cosmosResponseMessage.Headers,
+                    item);
+            });
+        }
+
+        internal Task<CosmosStoredProcedureResponse> CreateStoredProcedureResponse(Task<CosmosResponseMessage> cosmosResponseMessageTask)
```
Comment (simplynaveen20) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosResponseFactory.cs]: Thought of moving these scripts create factory methods to extension methods and moving them to script name space ?
That way everything of script will be under one namespace

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosResponseFactory.cs]: I would prefer if we stayed with the current implementation and avoid extension methods. This is an internal class so it won't impact users at all. Extension methods should really only be used for sealed classes. They break the basic object oriented design principals which at least in my experience makes development very painful.

File: Microsoft.Azure.Cosmos/src/Resource/Settings/CosmosTriggerSettings.cs [hunk 2]
```diff
@@ -16,21 +15,14 @@ namespace Microsoft.Azure.Cosmos
     /// Azure Cosmos DB supports pre and post triggers written in JavaScript to be executed on creates, updates and deletes. 
     /// For additional details, refer to the server-side JavaScript API documentation.
     /// </remarks>
-    internal class CosmosTriggerSettings
+    public class CosmosTriggerSettings
```
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/CosmosTriggerSettings.cs]: All the other `...Settings` classes are not `sealed`. Shouldn't this be consistent with the rest?

File: Microsoft.Azure.Cosmos/src/Resource/Settings/TriggerOperation.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
+﻿//------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos
+{
+    /// <summary>
+    /// Specifies the operations on which a trigger should be executed in the Azure Cosmos DB service.
+    /// </summary> 
+    public enum TriggerOperation : short
```
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/TriggerOperation.cs]: This is currently mapping an enum in Direct, Direct is singular.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/GatewayTests.cs [hunk 4]
```diff
@@ -418,7 +419,7 @@ internal void ValidateTriggerCrud(ConsistencyLevel consistencyLevel, ConnectionP
                 Id = triggerName,
                 Body = "function() {var x = 10;}",
                 TriggerType = Documents.TriggerType.Pre,
-                TriggerOperation = TriggerOperation.All
+                TriggerOperation = Documents.TriggerOperation.All
```
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/GatewayTests.cs]: This is V2 tests

[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/GatewayTests.cs]: V3 tests are updated in this PR, see `TriggersTests.cs`

File: Microsoft.Azure.Cosmos/src/Scripts/CosmosScriptsCore.cs [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+    using System;
+    using System.Diagnostics;
+    using System.IO;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Documents;
+
+    internal sealed class CosmosScriptsCore : CosmosScripts
+    {
+        private readonly CosmosContainerCore container;
```
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Scripts/CosmosScriptsCore.cs]: Done

File: Microsoft.Azure.Cosmos/src/Scripts/CosmosScriptsCore.cs [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+            Task<CosmosResponseMessage> response = this.ProcessStreamOperationAsync(
+                resourceUri: LinkUri,
+                resourceType: ResourceType.StoredProcedure,
+                operationType: OperationType.Replace,
+                requestOptions: requestOptions,
+                partitionKey: null,
+                streamPayload: CosmosResource.ToStream(storedProcedureSettings),
+                cancellation: cancellation);
+
+            return this.container.ClientContext.ResponseFactory.CreateStoredProcedureResponse(response);
```
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Scripts/CosmosScriptsCore.cs]: Refactored

File: Microsoft.Azure.Cosmos/src/Scripts/CosmosScriptsCore.cs [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+            {
+                throw new ArgumentNullException(nameof(id));
+            }
+
+            if (string.IsNullOrEmpty(body))
+            {
+                throw new ArgumentNullException(nameof(body));
+            }
+
+            CosmosStoredProcedureSettings storedProcedureSettings = new CosmosStoredProcedureSettings()
```
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Scripts/CosmosScriptsCore.cs]: Added an internal constructor

File: Microsoft.Azure.Cosmos/src/Resource/Settings/CosmosTriggerSettings.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/CosmosTriggerSettings.cs]: Need some guidance here. Not in scope for this PR.

**Comment by kirankumarkolli** [comment]
Please create an issue (Breaking tag) with details and map to this PR.
It will help us on release notes tracking as well.

**Comment by j82w** [comment]
Can you please update the description with the latest design changes?

File: Microsoft.Azure.Cosmos/src/Scripts/CosmosScripts.cs [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+        ///         body: sprocBody);
+        /// 
+        /// // Execute the stored procedure
+        /// CosmosItemResponse<string> sprocResponse = await scripts.ExecuteStoredProcedureAsync<string, string>(testPartitionId, "appendString", "Item as a string: ");
+        /// Console.WriteLine("sprocResponse.Resource");
+        /// ]]>
+        /// </code>
+        /// </example>
+        public abstract Task<StoredProcedureResponse> CreateStoredProcedureAsync(
+                    string id,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Scripts/CosmosScripts.cs]: Lets use settings are argument

File: Microsoft.Azure.Cosmos/src/Scripts/CosmosScripts.cs [hunk 9]
```diff
@@ -0,0 +1,652 @@
...
+        ///     getContext().getRequest().setBody(item);
+        /// }";
+        /// 
+        /// CosmosScripts scripts = this.container.GetScripts();
+        /// CosmosResponseMessage response = await scripts.ReplaceStoredProcedureAsync("testTriggerId", body);
+        /// ]]>
+        /// </code>
+        /// </example>
+        public abstract Task<StoredProcedureResponse> ReplaceStoredProcedureAsync(
+            string id,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Scripts/CosmosScripts.cs]: settings as argument.

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Scripts/CosmosScripts.cs]: Done

File: Microsoft.Azure.Cosmos/src/Scripts/CosmosScripts.cs (hunk 8 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Scripts/CosmosScripts.cs]: Done

**Pull request merged**